File: widgets/workspaces.cpp
#include "include/widgets/workspaces.hpp"

namespace widgets {

size_t workspaces_count = 10;

Workspaces::Workspaces(io::Ctx *ctx)
    : Gtk::Box(), utils::Subscriber(ctx), buttons(workspaces_count) {
  set_orientation(Gtk::Orientation::HORIZONTAL);
  set_spacing(0);
  set_css_classes({"widget", "workspaces"});
  set_name("Workspaces");

  for (size_t idx = 0; idx < workspaces_count; idx++) {
    auto &button = buttons.at(idx);
    button.set_label(std::to_string(idx + 1));
    append(button);

    button.signal_clicked().connect(
        [ctx, idx]() { io::io_hyprland_go_to_workspace(idx, ctx); });
  }
}

void Workspaces::on_io_event(io::Event::Workspaces_Body data) {
  for (size_t idx = 1; idx <= workspaces_count; idx++) {
    auto &button = buttons.at(idx - 1);
    auto visible = false;
    for (size_t i = 0; i < data.ids.len; i++) {
      if (data.ids.ptr[i] == idx) {
        visible = true;
      }
    }
    button.set_visible(visible || idx <= 5);
    button.remove_css_class("active");
    button.remove_css_class("inactive");
    if (idx == data.active_id) {
      button.add_css_class("active");
    } else {
      button.add_css_class("inactive");
    }
  }
}

} // namespace widgets

File: ui/top_bar/workspaces.py
from gi.repository import Gtk
from utils.commands import Commands
from utils.context import ctx


class Workspaces(Gtk.Box):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        ctx.pub_sub.subscribe(self)
        self.workspaces_count = 10

        self.set_orientation(Gtk.Orientation.HORIZONTAL)
        self.set_spacing(0)
        self.set_css_classes(["widget", "workspaces"])
        self.set_name("Workspaces")

        self.buttons = []
        for idx in range(0, self.workspaces_count):
            button = Gtk.Button(label=f"{idx + 1}")
            self.append(button)
            self.buttons.append(button)
            button.connect("clicked", lambda _, idx=idx: self.on_click(idx))

    def on_workspaces(self, event):
        for idx in range(0, self.workspaces_count):
            button = self.buttons[idx]
            visible = (idx + 1) in event.ids
            button.set_visible(visible or idx < 5)
            if idx + 1 == event.active_id:
                button.set_css_classes(["active"])
            else:
                button.set_css_classes([])

    def on_click(self, idx):
        Commands.hyprland_go_to_workspace(idx)
